### PR TITLE
Fixes to enable view fusion in LTC

### LINKED ISSUE #1366: Nested View Handling
Currently we throw an error when we encounter a nested view.

Options:
1. Graph-Level - Separate nested views in graph_fuser.cpp (single-kernel)
2. Segmentation-Level - Separate nested views into different kernels

```
TEST_F(NVFuserTest, FusionViewNested_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  std::vector<int64_t> input_shape{256, 128, 4096};
  std::vector<int64_t> output_shape{32768, 4096};

  TensorView* x = makeSymbolicTensor(input_shape.size());
  TensorView* bias = makeSymbolicTensor(input_shape.size());
  fusion.addInput(x);
  fusion.addInput(bias);

  auto x_add_bias = add(x, bias);
  auto x_view = view(x_add_bias, input_shape, output_shape);
  auto x_view_revert = view(x_view, output_shape, input_shape);
  auto x_gelu = gelu(x_view_revert);
  fusion.addOutput(x_gelu);
  fusion.addOutput(x_view_revert);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor at_x = at::randn(input_shape, options);
  at::Tensor at_bias = at::randn(input_shape, options);
  std::vector<IValue> aten_inputs = {at_x, at_bias};

  auto lparams = schedulePointwise(&fusion, aten_inputs);

  FusionExecutor fe;
  fe.compileFusion(&fusion);
  auto outputs = fe.runFusion(aten_inputs, lparams);

  auto at_x_add_bias = at_x + at_bias;
  auto at_x_view = at::native::view(at_x_add_bias, output_shape);
  auto at_x_view_revert = at::native::view(at_x_view, input_shape);
  auto at_gelu = at::gelu(at_x_view_revert, false);

  testValidate(&fusion, outputs, aten_inputs, {at_x_view_revert, at_gelu}, __LINE__, __FILE__);
}
```

### PR DESCRIPTION
**Pull request by rdspring1**
- Allows `bias+view+gelu` fusion create by linear layer in LTC
- Handles nested view / reshape in graph-fuser.cpp
- Fixes #1366, #1367

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -3154,7 +3154,7 @@ def t_bias(x: torch.Tensor, w: torch.Tensor, bias: torch.Tensor):
             jit_o = jitted_bias(inp, weight, bias)
 
         graph = jitted_bias.graph_for(inp)
-        self.assertGraphContainsExactly(graph, FUSION_GROUP, 0)
+        self.assertGraphContains(graph, FUSION_GROUP, True)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:3227]: hmmm, are we having multiple fusion here because of the new view op?

File: test/test_jit_cuda_fuser.py [hunk 2]
```diff
@@ -3374,27 +3374,28 @@ def __init__(self):
                 with torch.no_grad():
                     self.bias.fill_(10)
 
-            def forward(self, inputs : torch.Tensor, view_shape : List[int]):
+            def forward(self, inputs : torch.Tensor, bias : torch.Tensor, view_shape : List[int]):
                 o = inputs.view(view_shape)
-                inputs = inputs * self.bias
+                inputs.add_(bias)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:3449]: Just curious: any reason we want to change this to an inplace update?

File: test/test_jit_cuda_fuser.py [hunk 3]
```diff
@@ -3536,7 +3572,7 @@ def forward(self, inputs : torch.Tensor, bias : torch.Tensor):
 
         self.assertEqual(o.dtype, jit_o.dtype)
         self.assertTrue(self._compare("comparing output failed", o, jit_o, error))
-        graph = t_jit.graph_for(x)
+        graph = t_jit.graph_for(x, bias)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:3645]: errr... how was this test passing before the change?! bias doesn't seem to be an optional argument.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 4]
```diff
@@ -1209,7 +1209,12 @@ struct CudaGraphFuser {
 
     for (Node* node : block_->nodes()) {
       for (Block* sub_block : node->blocks()) {
-        CudaGraphFuser(sub_block, graph_).run();
+        CudaGraphFuser sub_block_cfg(sub_block, graph_);
+        sub_block_cfg.run();
+        // Accumulate runtime shapes for all sub-blocks
+        fusion_value_to_runtime_shape_.insert(
+            sub_block_cfg.fusion_value_to_runtime_shape_.begin(),
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1216]: This one looks promising... I am currently seeing an error `RuntimeError: fusion_value_to_runtime_size.find(self_value) != fusion_value_to_runtime_size.end()`
Hopefully it'll fix that.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 5]
```diff
@@ -1566,6 +1571,7 @@ void guardFusionGroup(
 
       // step a. handle fusion
       // remove inputs to fusion, and update check logic for fallback
+      auto profiled_ival_output = fusion->input(offset);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp]: I don't see `profiled_ival_output` used anywhere.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 6]
```diff
@@ -2233,6 +2254,67 @@ bool removeInplaceOperations(const std::shared_ptr<Graph>& graph) {
       graph, [&](Node* node) { return inplace_ops.count(node->kind()) != 0; });
 }
 
+// Recursively traverse blocks, gather all nodes with given symbol,
+// and then apply mutator function.
+void mutateNode(
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:2241]: nitpick: this seems to be a general version of `traverseProfileIValues`. can we merge the two?

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 7]
```diff
@@ -2474,6 +2474,8 @@ class IrParser {
 
     {
       std::array<const char*, kNumViewOps> ViewOps = {
+          "aten::reshape(Tensor(a) self, int[] shape) -> Tensor(a)",
+          "aten::view(Tensor(a) self, int[] size) -> Tensor(a)",
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp]: I thought that we are not fusing `aten::view` since it's not always safe to do so?

**Review by jjsjann123 — COMMENTED** [review]
looks like we are now fusing `aten::view` as well as `aten::view_copy`.

Wasn't it unsafe to fuse random `aten::view`? An in this instance, there's no point of converting `view_copy` any more.

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py:3227]: I am seeing a `view + bias` fusion here.

File: test/test_jit_cuda_fuser.py (hunk 2 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py:3449]: It wasn't creating an aliased operation.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp (hunk 5 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp]: I must have missed removing it.

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 7 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: We shouldn't fuse `aten::view` in `CudaFusionGroup`. If it wasn't included in `parser.cpp`, then the shape information is not propagated through the graph.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: I thought shape propagation is only done inside graph fusion, so it should have happened after we construct a fusion node within its subgraph, which should only have `view_copy/reshape_copy`. Curious to see how/why we would need to add `reshape/view` into our parser.

On a separate note, putting reshape/view here means we are fusing them anyway. So this would break the alias check we have.

File: test/test_jit_cuda_fuser.py (hunk 3 above)
Comment (rdspring1) [reply test/test_jit_cuda_fuser.py:3645]: I was surprised why the test didn't complain. 🤷🏼‍♂️

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 7 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/parser.cpp]: I think #1443 fixed the shape propagation issue.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 8]
```diff
@@ -2077,47 +2068,55 @@ void decomposeLinearOps(Block* block) {
...
     }
   }
 
-  auto alias_db = std::make_unique<AliasDb>(graph);
-  for (Node* n : maybe_alias_nodes) {
-    if (!alias_db->safeToChangeAliasingRelationship(
-            n->input(0), n->output(0))) {
-      continue;
-    }
+  auto alias_db = std::make_shared<AliasDb>(graph);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp]: nitpick, don't see we sharing alias_db around, any reason that we changed it from unique_ptr to shared_ptr?

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp [hunk 9]
```diff
@@ -105,6 +106,9 @@ static const OperatorMap<std::string>& get_schema_to_function_graph() {
       {"aten::cat(Tensor[] tensors, int dim=0) -> Tensor", "cat"},
       {"aten::permute(Tensor(a) self, int[] dims) -> Tensor(a)", "permute"},
       {"aten::view(Tensor(a) self, int[] size) -> Tensor(a)", "view"},
+      {"aten::reshape(Tensor(a) self, int[] shape) -> Tensor(a)", "view"},
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_shape_registry.cpp:109]: I don't know to what extend we'll need these.... (Elias pointed to me earlier that some of these ops are handled in a different file, but we can clean it up in the next upstream PR)

**Review by jjsjann123 — APPROVED** [review]
LGTM and also gets rid of the failures on opinfo test I've been fighting!!!
:tada:

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 10]
```diff
@@ -788,7 +788,12 @@ TensorView* broadcast(
           ParallelType::Serial,
           IterType::BroadcastWithoutStride));
     } else {
-      out_domain.push_back(inp_domain[iinp]->clone());
+      out_domain.push_back(IrBuilder::create<IterDomain>(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.cpp:821]: Do we need to more carefully audit `IterDomain::clone` usage? Should rfactor flag be an optional argument in clone?

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp [hunk 11]
```diff
@@ -445,8 +447,8 @@ void ComputeAtMap::build(Fusion* fusion, GpuLower* gpu_lower) {
       int concrete_count = n_concrete_ids_.at(id);
       if (concrete_count >= max_concrete_count) {
         int broadcast_count = n_broadcast_ids_.at(id);
-        if (concrete_count > max_concrete_count ||
-            broadcast_count > max_broadcast_count) {
+        if (concrete_count >= max_concrete_count ||
+            broadcast_count >= max_broadcast_count) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: Why is this necessary?

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp [hunk 12]
```diff
@@ -1170,14 +1170,24 @@ std::unique_ptr<Fusion> SegmentedFusion::makeFusion(SegmentedGroup* sg) {
...
+      view_tvs.push_back(clone_tv->as<TensorView>());
+    }
   }
 
   for (auto out : getAllOutputs(sg)) {
     fusion_segment->addOutput(complete_to_segment_map.clone(out));
   }
 
+  for (auto tv : view_tvs) {
+    tv->convertRfactorToRootDomain();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1188]: Is this still a problem after you moved all those instances of `getRootDomain` to `getMaybeRfactorDomain`?

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp [hunk 13]
```diff
@@ -193,6 +193,16 @@ bool Val::isOneInt() const {
   return int_val.has_value() && int_val.value() == 1;
 }
 
+bool Val::isDefinition(ExprType expression_type) const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.cpp]: Nit: `Val::isDefinitionType`

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp [hunk 14]
```diff
@@ -432,8 +432,9 @@ class DomainMap {
     }
 
     // Erase all input concrete IDs mapped to the output domain
+    // Ignore unresolved broadcast dimensions
     for (auto out_id : output_tv->getMaybeRFactorDomain()) {
-      if (!out_id->isBroadcast() && !out_id->isReduction()) {
+      if (!out_id->isBroadcast()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:437]: Was the check on reduction harmful or just unnecessary?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 15]
```diff
@@ -119,6 +119,73 @@ TensorView::TensorView(
...
+      new_root_domain[idx] = IrBuilder::create<IterDomain>(
+          container(),
+          id->start(),
+          replacement_extents[idx],
+          id->stopOffset(),
+          id->getParallelType(),
+          id->getIterType());
+      ++idx;
+    } else {
+      new_root_domain[idx++] = id->clone();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:143]: Don't you want to clear the rfactor flag here?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 16]
```diff
@@ -119,6 +119,73 @@ TensorView::TensorView(
...
+    for (auto id : tv->getMaybeRFactorDomain()) {
+      if (!id->extent()->isConstScalar()) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  const auto kThisIsConcreteTensor = is_concrete_tensor(this);
+  // Create new symbolic extents for rfactor iterDomains
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Why is there special handling on if concrete? Can't you simply make new ID's for the entire domain?

File: torch/csrc/jit/codegen/cuda/transform_iter.cpp [hunk 17]
```diff
@@ -353,12 +361,20 @@ BestEffortReplay::BestEffortReplay(
...
+      TORCH_INTERNAL_ASSERT(
+          no_missing_exprs || any_target_expr_contains_broadcast_id, err_str);
     }
 
     // If any inputs are missing, continue as this expr doesn't match.
     if (missing_replay_input) {
-      TORCH_INTERNAL_ASSERT(!replay_has_rfactor_inp, err_str);
+      TORCH_INTERNAL_ASSERT(
+          !replay_has_rfactor_inp || any_target_expr_contains_broadcast_id,
+          err_str);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/transform_iter.cpp:377]: Note to understand this better. Don't recall what BestEffortReplay is doing here.

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h [hunk 18]
```diff
@@ -209,6 +209,13 @@ class TORCH_CUDA_CU_API TensorView : public Val {
     return domain_;
   }
 
+  //! This is for a TensorView with an rFactor domain that is an input to a
+  //! fusion segment. We convert the rfactor domain into a new root domain.
+  //! Any dynamic-sized rfactor iterDomains are given a new symbolic extent.
+  //! Concrete integer extents are kept. Output TensorViews of any subsequent
+  //! expressions that use this TensorView are also updated.
+  void convertRfactorToRootDomain();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:217]: These functions seem fine, but again wondering if rfactor domain is consistently used why this is needed.

File: torch/csrc/jit/runtime/symbolic_shape_registry.cpp (hunk 9 above)
Comment (rdspring1) [reply torch/csrc/jit/runtime/symbolic_shape_registry.cpp:109]: I should remove/update the aten::gelu_backward entry.

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 11 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: I encounter this exception without the change.
```
C++ exception with description "index_map.find(root_dom[i]) != index_map.end()
INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":2167, please report a bug to PyTorch. 
Couldn't find root mapping for TV6 dim: 0 id: iS143{T1.size[0]}
Exception raised from getNonGlobalConsumerStridedIndices at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:2167
```

Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: IIUC, the concrete ID of the reference tensor is missing a mapping to a root domain in the consumer tensor.

When selecting the concrete ID, the id with the missing root domain has the same `concrete_count` and `broadcast_count`, so it is not selected.

I think this is related to moving the rfactor tensor to a root tensor. The new root ID is counted as a single domain instead of being derived from multiple root domains.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 10 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/arith.cpp:821]: I think we should avoid propagating the rfactor flag if it isn't in an rfactor domain.

File: torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp (hunk 14 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:437]: It prevented the pointwise scheduler from handling trivial reduction operations.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 15 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:143]: The rfactor flag should be false because we only create a new extent for rfactor iterDomains.

I'll add an assertion.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 16 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: I only wanted to make new IDs for rfactor iterDomains. The remaining domains can be cloned normally. Special handling for concrete tensors shouldn't be necessary.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp (hunk 12 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:1188]: Yes, I was seeing an illegal memory access because the fusion input predicate was for the root domain instead of the rfactor domain.

It seems there is an implicit assumption that the fusion input does not have a rfactor domain.

**Review by rdspring1 — COMMENTED** [review]
My approach in this PR is to convert the rfactor domain into a stand-alone root domain when we segment on a view tensor.

I'll create an issue for handling fusion inputs with a rfactor domain.

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 11 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: @naoyam could you take a look at this please? This seems odd to me.

**Comment by naoyam** [comment]
Why not moving all view-related tests to their own source file?

File: torch/csrc/jit/codegen/cuda/compute_at_map.cpp (hunk 11 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: FYI, the ViewReductionShmoo test causes the issue.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/compute_at_map.cpp]: This needs to be resolved with the code from with https://github.com/csarofeen/pytorch/pull/1494

**Review by csarofeen — APPROVED** [review]
LGTM, just one outstanding comment on the code that should be reverted/resolved in compute at map with https://github.com/csarofeen/pytorch/pull/1494
